lib/debugger-wrapper.js

```diff
@@ -47,27 +47,38 @@ function init() {
 
 function instrumentScript(scriptUrl) {
   const moduleInfo = moduleUtils.getModuleInfo(scriptUrl);
-  const vulnerableMethods = snapshot.getVulnerableFunctionsLocations(moduleInfo);
-  Object.keys(vulnerableMethods).forEach((methodName) => {
-    const methodLocation = vulnerableMethods[methodName];
-    transmitter.addInstrumentedFunction({url: scriptUrl, moduleInfo, methodName, methodLocation});
-    setBreakpointOnFunction(scriptUrl, moduleInfo, methodName, methodLocation);
+  const functionsToInstrument = snapshot.getVulnerableFunctionsLocations(moduleInfo);
+
+  Object.keys(functionsToInstrument).forEach((functionName) => {
+    const functionLocation = functionsToInstrument[functionName];
+    if (!(functionName in scriptUrlToInstrumentedFunctions[scriptUrl])) {
+      setBreakpointOnFunction(scriptUrl, moduleInfo, functionName, functionLocation);
+    }
+  });
+
+  Object.keys(scriptUrlToInstrumentedFunctions[scriptUrl]).forEach((functionName) => {
```

Comment:
When could this ever be hit? This method is only ever called immediately after setting `scriptUrlToInstrumentedFunctions[scriptUrl]` to `{}`.

Or are we expecting this function to be called later on some new timer?

I feel it would be neater to extend the suspension logic to just not re-add a suspended breakpoint which is no longer in the list.

Reply:
currently, never (or only in `debugger.test.js` where it's white-box tested).
the next phase will add some sort of timer to retrieve the latest snapshot, and then iterate over `scriptUrlToInstrumentedFunctions` and invoke `instrumentScript` for each script there, to add missing bps / remove obsolete ones.